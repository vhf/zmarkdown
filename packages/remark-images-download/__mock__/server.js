import path from 'path'
import express from 'express'

const app = express()
app.use('/', express.static(path.join(__dirname, 'files')))

const server = app.listen(27273, () => {})

module.exports = server
