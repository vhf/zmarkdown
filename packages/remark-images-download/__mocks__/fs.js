const fs = jest.genMockFromModule('fs')

fs.stat = (path, callback) => {
  const err = null
  const stats = {
    isDirectory () {
      return true
    }
  }
  process.nextTick(() => callback(err, stats))
}

fs.mkdir = (path, callback) => {
  process.nextTick(() => callback())
}

module.exports = fs
