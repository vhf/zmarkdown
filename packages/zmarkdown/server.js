const toVFile = require('to-vfile')
const clone = require('clone')

const remarkImagesDownload = require('remark-images-download/src')

const rebberStringify = require('rebber/src')
const rebberConfig = require('./config/rebber')
const zmarkdown = require('./common')

const fromFile = (filepath) => toVFile.readSync(filepath)

const remarkConfig = require('./config/remark')
let zmd

function getLatexProcessor (config) {
  config.remarkConfig.noTypography = true

  return zmd.zmdParser(config.remarkConfig, config.extraPlugins)
    .use(rebberStringify, config.rebberConfig)
}

function rendererFactory (config, target) {
  if (target === 'html') {
    return zmd.rendererFactory(config)
  } else if (target === 'latex') {
    return zmd.rendererFactory(config, getLatexProcessor)
  }
  throw new Error(`Unknown target: ${target}`)
}

module.exports = (
  opts = {remarkConfig, rebberConfig},
  target = 'html'
) => {
  if (!opts.remarkConfig || !Object.keys(remarkConfig).length) {
    opts.remarkConfig = clone(remarkConfig)
  }

  if ((target === 'latex') && (!opts.remarkConfig || !Object.keys(remarkConfig).length)) {
    opts.rebberConfig = clone(rebberConfig)
  }

  if (!opts.extraPlugins) {
    opts.extraPlugins = [
      {obj: remarkImagesDownload, option: opts.remarkConfig.imagesDownload},
    ]
  }

  if (target === 'latex') {
    opts.remarkConfig.canUseTextr = false
  }

  zmd = zmarkdown(opts)
  zmd.renderString = rendererFactory(opts, target)
  zmd.renderFile = (path, cb) => {
    return rendererFactory(opts, target)(fromFile(path), cb)
  }

  return zmd
}
