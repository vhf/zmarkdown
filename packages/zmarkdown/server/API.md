# zmarkdown server HTTP API

### Request
All endpoints respond to `HTTP POST` requests with these keys in their body:

* `md` - required, markdown source string
* `opts` - optional, JSON options

`md` being a boring Markdown string, only `opts` will be specified below.

### Response

All endpoints return `[contents, metadata]` as JSON.

# Endpoints

## `/html` - Markdown to HTML

### Request

* `opts.disable_ping`, bool, default: `false`

  [pings][ping] won't get parsed

* `opts.disable_jsfiddle`, bool, default: `false`

  JSFiddle [iframes][iframes] are disabled

* `opts.inline`, bool, default: `false`

  Only parse inline Markdown elements (such as links and emphasis, unlike lists and fenced code blocks)

### Response

## `/latex` - Markdown to latex

### Request

* `opts.disable_jsfiddle`, bool, default: `false`

  see `/html`

### Response

## `/latex-document` - Markdown to tex file

### Request

* `opts.disable_jsfiddle`, bool, default: `false`

  see `/html`

* `opts.svg_dest`, string, default: <none> (SVGs won't get saved to file)

  Absolute path to the directory where Mairmaid's SVGs will be stored.

### Response

[ping]: https://www.npmjs.com/package/remark-ping
[iframes]: https://www.npmjs.com/package/remark-iframes
