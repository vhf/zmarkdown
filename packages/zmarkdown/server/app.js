/* eslint-disable no-console */
const express = require('express')
const bodyParser = require('body-parser')
const cors = require('cors')
const path = require('path')

const zmdVersion = require('../package.json').version

function controllerFactory (handler) {
  return (req, res) => {
    handler(req.body.md, req.body.opts, (err, result) => {
      if (err) {
        Raven.captureException(err, {req})
        console.error(err)
        res.status(500).json(null)
        return
      }

      res.json(result)
    })
  }
}

const Raven = require('raven')
Raven.config(process.env.SENTRY_DSN, {
  release: process.env.SENTRY_RELEASE || zmdVersion,
  environment: process.env.SENTRY_ENVIRONMENT || process.env.ZDS_ENVIRONMENT
}).install()

const {
  toHTML,
  toLatex,
  toLatexDocument,
} = require('./handlers')(Raven)

const app = express()

app.use(cors())

app.use(Raven.requestHandler())
app.use(Raven.errorHandler())

app.use('/static', express.static(path.join(__dirname, 'static')))

app.use(bodyParser.json({limit: '5mb'}))
app.post('/latex', controllerFactory(toLatex))
app.post('/latex-document', controllerFactory(toLatexDocument))
app.post('/html', controllerFactory(toHTML))

module.exports = app
