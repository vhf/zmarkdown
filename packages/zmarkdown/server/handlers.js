const clone = require('clone')
const pmx = require('pmx')
const zmarkdown = require('../common')
const remarkConfig = require('../config/remark')
const rebberConfig = require('../config/rebber')

const remarkImagesDownload = require('remark-images-download/src')

const probe = pmx.probe()

const endpoints = ['toEPUB', 'toHTML', 'toLatex', 'toLatexDocument']

const meters = endpoints.reduce((acc, endpoint) => {
  const meter = probe.meter({
    name: `${endpoint} rpm`,
    samples: 1,
    timeframe: 60,
  })

  const counter = probe.counter({
    name: `${endpoint} counter`,
    agg_type: 'sum',
  })

  acc[endpoint] = () => {
    meter.mark()
    counter.inc()
  }
  return acc
}, {})

// this object is used to memoize confirenderFilegured processors
const processors = {}

module.exports = function markdownHandlers (Raven) {
  return {
    toEPUB,
    toHTML,
    toLatex,
    toLatexDocument,
  }

  function toEPUB (markdown, opts = {}, callback) {
    meters.toEPUB()
    const target = 'html'

    opts.heading_shift = 2
    opts.disable_ping = true
    opts.disable_jsfiddle = true
    opts.inline = false

    render(target, markdown, opts, callback)
  }

  function toHTML (markdown, opts = {}, callback) {
    meters.toHTML()
    const target = 'html'

    opts.heading_shift = 2
    opts.disable_images_download = true

    render(target, markdown, opts, callback)
  }

  function toLatex (markdown, opts = {}, callback) {
    meters.toLatex()
    const target = 'latex'

    opts.heading_shift = 0
    opts.disable_ping = true
    opts.disable_jsfiddle = true

    render(target, markdown, opts, callback)
  }

  function toLatexDocument (markdown, opts = {}, callback) {
    meters.toLatexDocument()
    const target = 'latex'
    const template = zmarkdown().latexDocumentTemplate

    opts.heading_shift = 0
    opts.disable_ping = true
    opts.disable_jsfiddle = true

    render(target, markdown, opts, (err, [latex, metadata, messages] = []) => {
      if (err) return callback(err, markdown)

      template(Object.assign(opts, {latex}), (err, latexDocument) => {
        if (err) return callback(err)
        callback(null, [latexDocument, {}, messages])
      })
    })
  }

  function render (target, markdown, opts = {}, callback) {
    if (typeof markdown !== 'string') {
      return callback(new Error(`Markdown to render should be 'string', got` +
      `'${typeof markdown}' instead`))
    }
    if (!['html', 'latex'].includes(target)) {
      return callback(new Error(`Unknown target 'target=${target}'`))
    }

    /* zmd parser memoization */
    const key = String(target) + JSON.stringify(opts)
    if (!processors.hasOwnProperty(key)) {
      const remark = clone(remarkConfig)
      const rebber = clone(rebberConfig)

      /* presets */
      if (opts.disable_ping === true) {
        remark.ping.pingUsername = () => false
      }

      if (typeof opts.heading_shift === 'number') {
        remark.headingShifter = opts.heading_shift
      }

      if (opts.disable_jsfiddle === true) {
        remark.iframes['jsfiddle.net'].disabled = true
        remark.iframes['www.jsfiddle.net'].disabled = true
      }

      remark.imagesDownload.disabled = opts.disable_images_download === true
      if (remark.imagesDownload.disabled !== true) {
        if (opts.images_download_dir) {
          remark.imagesDownload.downloadDestination = opts.images_download_dir
        }
        if (
          Array.isArray(opts.local_url_to_local_path) &&
          opts.local_url_to_local_path.length === 2
        ) {
          remark.imagesDownload.localUrlToLocalPath = opts.local_url_to_local_path
        }
      }

      if (opts.inline === true) {
        remark.disableTokenizers = {
          block: [
            'indentedCode',
            'fencedCode',
            'blockquote',
            'atxHeading',
            'setextHeading',
            'footnote',
            'table',
            'custom_blocks',
          ],
        }
      }

      processors[key] = zmarkdown({remarkConfig: remark,
        rebberConfig: rebber,
        extraPlugins: [
          {obj: remarkImagesDownload, option: remark.imagesDownload},
        ]}, target)
    }

    processors[key].renderString(markdown, (err, {contents, data, messages} = {}) => {
      const metadata = data

      if (err) {
        Raven.mergeContext({
          extra: {
            zmdConfig: makeSerializable(processors[key].config),
            markdown: markdown,
            zmdOutput: {
              contents: contents,
              metadata: metadata,
              messages: messages,
            },
          },
        })
        return callback(err, markdown)
      }

      callback(null, [contents, metadata, messages])
    })
  }
}

function makeSerializable (obj) {
  return JSON.parse(JSON.stringify(obj))
}
