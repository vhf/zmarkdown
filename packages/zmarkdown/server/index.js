/* eslint-disable no-console */

const app = require('./app')

const server = app.listen(27272, () => {
  const host = server.address().address
  const port = server.address().port

  console.warn('zmarkdown server listening at http://%s:%s', host, port)
})
