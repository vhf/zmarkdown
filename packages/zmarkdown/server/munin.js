const pm2 = require('pm2')
module.exports = (isConfig) => (req, res) => {
  const endpoints = ['toEPUB', 'toHTML', 'toLatex', 'toLatexDocument']

  const supportedStats = {
    status: {
      graph: [
        'graph_title Process Status',
        'graph_vlabel',
      ],
      fields: (name) => [
        `${name}.label ${name}`,
        `${name}.critical 1`,
      ],
    },
    memory: {
      stack: true,
      graph: [
        'graph_title RAM Usage',
        'graph_vlabel Bytes',
        'graph_args --base 1024 --lower-limit 0',
      ],
      fields: (name) => [
        `${name}.label ${name}`,
        `${name}.draw STACK`,
      ],
    },
    cpu: {
      graph: [
        'graph_title CPU Usage',
        'graph_vlabel percent',
        'graph_args --base 1000 --lower-limit 0 --rigid',
        'graph_scale no',
      ],
      fields: (name) => `${name}.label ${name}`,
    },
    event_loop_lag: {
      graph: [
        'graph_title Event Loop Lag',
        'graph_vlabel ms',
        'graph_args --lower-limit 0',
      ],
      fields: (name) => `${name}.label ${name}`,
    },
    req_per_process: {
      stack: true,
      graph: [
        'graph_title Request Per Process',
        'graph_vlabel requests',
        'graph_args --lower-limit 0',
      ],
      fields: (name) => [
        `${name}.label ${name}`,
        `${name}.draw STACK`,
      ],
    },
    avg_per_process: {
      stack: true,
      graph: [
        'graph_title Requests Per Second Per Process',
        'graph_vlabel requests per second',
        'graph_args --lower-limit 0',
      ],
      fields: (name) => [
        `${name}.label ${name}`,
        `${name}.draw STACK`,
      ],
    },
    req_per_endpoint: {
      stack: true,
      graph: [
        'graph_title Request Per Endpoint',
        'graph_vlabel requests',
        'graph_args --lower-limit 0',
      ],
      fields: () => endpoints.map(endpoint =>
        `${endpoint}.label ${endpoint}\n${endpoint}.draw STACK`),
    },
    avg_per_endpoint: {
      stack: true,
      graph: [
        'graph_title Requests Per Second Per Endpoint',
        'graph_vlabel requests per second',
        'graph_args --lower-limit 0',
      ],
      fields: () => endpoints.map(endpoint =>
        `${endpoint}.label ${endpoint}\n${endpoint}.draw STACK`),
    },
  }

  const status = {
    online: 0,
    stopped: 1,
    errored: 2,
  }

  const stat = req.params.plugin
  const collected = []

  if (!(supportedStats.hasOwnProperty(stat))) {
    return res.status(500).send({
      error: `Invalid stat, given : ${stat}, ` +
        `expected one of ${JSON.stringify(supportedStats, null, 2)}.`,
    })
  }

  if (isConfig) {
    run(printConfig)
  } else {
    run(printStats)
  }

  function gatherData (procList) {
    let maxMem = 0

    const procs = procList.reduce((acc, proc, i) => {
      let avgPerProcess
      let reqPerProcess
      let loopLag

      if (proc.pm2_env.axm_monitor) {
        reqPerProcess = endpoints.map(endpoint =>
          parseInt(proc.pm2_env.axm_monitor[`${endpoint} counter`].value)
        ).reduce((sum, val) => sum + val, 0)
        avgPerProcess = endpoints.map(endpoint =>
          parseFloat(proc.pm2_env.axm_monitor[`${endpoint} rpm`].value)
        ).reduce((sum, val) => sum + val, 0)

        loopLag = parseFloat(proc.pm2_env.axm_monitor['Loop delay'].value)
      }

      const data = {
        status: proc.pm2_env.status in status ? status[proc.pm2_env.status] : 3,
        memory: proc.monit.memory,
        cpu: proc.monit.cpu,
        event_loop_lag: loopLag,
        req_per_process: reqPerProcess,
        avg_per_process: avgPerProcess,
      }

      if (!maxMem && proc.pm2_env.max_memory_restart) {
        maxMem = proc.pm2_env.max_memory_restart
      }
      if (maxMem) data.maxMem = maxMem

      acc[`${proc.name}-${proc.pm_id}`] = data

      return acc
    }, {})

    const _endpoints = {
      req_per_endpoint: endpoints.reduce((acc, endpoint) => {
        acc[endpoint] = procList.reduce((sum, proc) =>
          sum + parseInt(proc.pm2_env.axm_monitor[`${endpoint} counter`].value), 0)
        return acc
      }, {}),

      avg_per_endpoint: endpoints.reduce((acc, endpoint) => {
        acc[endpoint] = procList.reduce((sum, proc) =>
          sum + parseFloat(proc.pm2_env.axm_monitor[`${endpoint} rpm`].value), 0)
        return acc
      }, {}),
    }

    return {procs, endpoints: _endpoints}
  }


  // prints a munin readable config
  function printConfig (stats) {
    if (!supportedStats[stat]) {
      return res.status(500).send({
        error: `"${stat}" not configured`,
      })
    }
    const procs = stats.procs

    l(supportedStats[stat].graph.join('\n'))
    l('graph_category zmd')

    let output = ''
    if (['req_per_endpoint', 'avg_per_endpoint'].includes(stat)) {
      output = supportedStats[stat].fields().join('\n')
    } else {
      output = Object.keys(procs).map(job => {
        const tmp = supportedStats[stat].fields(job)
        return Array.isArray(tmp)
          ? tmp.join('\n')
          : tmp
      }).join('\n')
    }
    if (supportedStats[stat].stack) {
      output = output.replace(' STACK', ' AREA')
    }
    l(output)

    pm2.disconnect()
    res.send(collected.join('\n'))
  }

  // prints a munin readable output
  function printStats (stats) {
    const procs = stats.procs

    if (['req_per_endpoint', 'avg_per_endpoint'].includes(stat)) {
      endpoints.forEach(endpoint => {
        l(`${endpoint}.value ${stats.endpoints[stat][endpoint]}`)
      })
    } else {
      Object.keys(procs).forEach((procName) => {
        const value = procs[procName][stat]
        l(`${procName}.value ${value}`)
      })
    }

    pm2.disconnect()
    res.send(collected.join('\n'))
  }

  function run (callback) {
    pm2.connect((err) => {
      if (err) {
        return res.status(500).send({error: err})
      }

      pm2.list((err, plist) => {
        if (err) {
          return res.status(500).send({error: err})
        }

        plist = plist.filter(process => !process.name.startsWith('pm2'))

        callback(gatherData(plist))
      })
    })
  }

  function l (string) {
    collected.push(string)
  }
}
