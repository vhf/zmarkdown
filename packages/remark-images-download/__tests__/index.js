/* global beforeAll, afterAll */
import clone from 'clone'
import fs from 'fs'
import dedent from 'dedent'
import path from 'path'
import rimraf from 'rimraf'

import unified from 'unified'
import reParse from 'remark-parse'
import stringify from 'rehype-stringify'
import remark2rehype from 'remark-rehype'

import plugin from '../src/'
import server from '../__mock__/server'

const DOWNLOAD_DESTINATION = '/tmp'
const downloadDestination = path.join(DOWNLOAD_DESTINATION, 'remark-image-download-tests')

const testFilesDir = `${__dirname.replace('__tests__', '__mock__')}/files`

const firstMsg = (vfile) => vfile.messages[0].message

const renderFactory = (opts = {}) =>
  (text) => unified()
    .use(reParse)
    .use(plugin, Object.assign(clone(opts), {downloadDestination}))
    .use(remark2rehype)
    .use(stringify)
    .process(text)

beforeAll((done) => {
  fs.mkdir(downloadDestination, done)
})

afterAll((done) => {
  server.close()
  rimraf(downloadDestination, done)
})

const r = (html) => html.replace(
  new RegExp(
    path.join(downloadDestination, '([a-zA-Z0-9_-]{7,14})', '([a-zA-Z0-9_-]{7,14})'),
    'g'),
  'foo/bar')

const noDownloads = async asyncFunction => {
  const prevDownloadDirCount = fs.readdirSync(downloadDestination).length

  await asyncFunction()

  const downloadDirCount = fs.readdirSync(downloadDestination).length
  expect(downloadDirCount).toBe(prevDownloadDirCount)
}

describe('mock server tests', () => {
  test('rejects invalid URLs', () => {
    const file = '![foo](http://%99:%99@example.com)'

    const render = renderFactory()

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toBe('Invalid URL: http://%99:%99@example.com')
    })
  })

  test('downloads image ok', () => {
    const file = dedent`
      ![](http://localhost:27273/test.svg)

      ![](http://localhost:27273/ok.png)

      ![](http://localhost:27273/ok.png?foo)
    `
    const html = dedent`
      <p><img src="foo/bar.svg"></p>
      <p><img src="foo/bar.png"></p>
      <p><img src="foo/bar.png"></p>
    `

    const render = renderFactory()

    return render(file).then(vfile => {
      expect(r(vfile.contents)).toBe(html)
      const downloadedFiles = fs.readdirSync(vfile.data.imageDir)
      expect(downloadedFiles.length).toBe(3)
    })
  })

  test('does not crash without images', () => {
    const file = `foo`
    const html = `<p>foo</p>`

    const render = renderFactory()

    return render(file).then(vfile => {
      expect(vfile.contents).toBe(html)
    })
  })

  test('does not create empty dest folder', async () => {
    const file = `foo`

    const render = renderFactory()

    return noDownloads(() =>
      render(file).then(vfile => {
        expect(vfile.data.imageDir).toBe(undefined)
      })
    )
  })

  test('skips bigger images and reports', () => {
    const file = `![](http://localhost:27273/ok.png)`
    const html = `<p><img src="http://localhost:27273/ok.png"></p>`

    const error = 'File at http://localhost:27273/ok.png weighs 516, max size is 400'

    const render = renderFactory({
      maxFileSize: 400,
    })

    return noDownloads(() =>
      render(file).then(vfile => {
        expect(firstMsg(vfile)).toBe(error)
        expect(vfile.contents).toBe(html)
        expect(vfile.data.imageDir).toBe(undefined)
      })
    )
  })

  test('default big images', () => {
    const defaultImagePath = 'default.png'
    const file = `![](http://localhost:27273/ok.png)`
    const html = `<p><img src="${downloadDestination}/${defaultImagePath}"></p>`

    const render = renderFactory({
      maxFileSize: 400,
      defaultImagePath,
      defaultOn: {
        fileTooBig: true,
      },
    })

    return render(file).then(vfile => {
      expect(vfile.contents).toBe(html)
    })
  })

  test('reports bad SVG', () => {
    const file = `![](http://localhost:27273/bad.svg)`
    const html = `<p><img src="http://localhost:27273/bad.svg"></p>`

    const render = renderFactory()

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toBe(
        'Could not detect http://localhost:27273/bad.svg mime type, not SVG either')
      expect(vfile.contents).toBe(html)
      expect(vfile.data.imageDir).toBe(undefined)
    })
  })

  test('skips wrong mimes', () => {
    const file = `![](http://localhost:27273/wrong-mime.txt)`
    const html = `<p><img src="http://localhost:27273/wrong-mime.txt"></p>`

    const render = renderFactory()

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toBe(
        'Content-Type of http://localhost:27273/wrong-mime.txt is not an image/ type')
      expect(vfile.contents).toBe(html)
    })
  })

  test('default wrong mime', () => {
    const defaultImagePath = 'default.png'
    const file = `![](http://localhost:27273/wrong-mime.txt)`
    const html = `<p><img src="${downloadDestination}/${defaultImagePath}"></p>`

    const render = renderFactory({
      defaultImagePath,
      defaultOn: {
        mimeType: true,
      },
    })

    return render(file).then(vfile => {
      expect(vfile.contents).toBe(html)
    })
  })

  test('reports bad mime headers', () => {
    const file = `![](http://localhost:27273/wrong-mime.txt)`

    const render = renderFactory()

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toBe(
        'Content-Type of http://localhost:27273/wrong-mime.txt is not an image/ type')
    })
  })

  test('reports bad mime content', () => {
    const file = `![](http://localhost:27273/wrong-mime.png)`

    const render = renderFactory()

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toBe(
        'Could not detect http://localhost:27273/wrong-mime.png mime type, not SVG either')
    })
  })

  test('reports protocol not whitelisted', () => {
    const file = `![](file://localhost:27273/wrong-mime.png)`

    const error = "Protocol 'file:' not allowed."

    const render = renderFactory()

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toBe(error)
    })
  })

  test('skips when directory reaches size limit', () => {
    const file = dedent`
      ![](http://localhost:27273/ok.png)
      ![](http://localhost:27273/ok.png?foo)
      ![](http://localhost:27273/ok.png?bar)
      ![](http://localhost:27273/ok.png?moo)
    `

    // query strings are required in order to make URLs unique (one single
    // file is downloaded otherwise).
    // each file being roughly 30% of directory size limit, the 4th one would exceed this
    // limit. It shouldn't get downloaded & replaced:

    const html = dedent`
      <p><img src="foo/bar.png">
      <img src="foo/bar.png">
      <img src="foo/bar.png">
      <img src="http://localhost:27273/ok.png?moo"></p>
    `

    const error = 'Cannot download http://localhost:27273/ok.png?moo because ' +
      'destination directory reached size limit'

    const render = renderFactory({
      maxFileSize: 516,
      dirSizeLimit: (516 * 3) + 50,
    })

    return render(file)
      .then(vfile => {
        expect(firstMsg(vfile)).toBe(error)
        expect(r(vfile.contents)).toBe(html)
      })
  })

  test('does not download when disabled', () => {
    const file = dedent`
      ![](http://localhost:27273/ok.png)
      ![](http://localhost:27273/ok.png)
      ![](http://localhost:27273/ok.png)
      ![](http://localhost:27273/ok.png)
    `

    // images will not be downloaded
    const html = dedent`
      <p><img src="http://localhost:27273/ok.png">
      <img src="http://localhost:27273/ok.png">
      <img src="http://localhost:27273/ok.png">
      <img src="http://localhost:27273/ok.png"></p>
    `

    const render = renderFactory({
      disabled: true,
    })

    return noDownloads(async () => {
      const vfile = await render(file)
      expect(vfile.contents).toBe(html)
    })
  })

  test('skips local images', () => {
    const file = `![](local.png)`
    const html = `<p><img src="local.png"></p>`

    const render = renderFactory()

    expect(render(file).then(vfile => vfile.contents)).resolves.toBe(html)
  })

  test('reports local magic number errors', () => {
    const file = `![](/foobar/wrong-mime.txt)`

    const render = renderFactory({
      localUrlToLocalPath: (localUrl) => {
        const localPath = __dirname.replace('__tests__', '__mock__')
        return `${localPath}/files${localUrl.slice(7)}`
      },
    })

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toMatch('mime type, not SVG either')
    })
  })

  test('copies local images with function', () => {
    const file = `![](/foobar/ok.png)`
    const html = `<p><img src="foo/bar.png"></p>`

    const render = renderFactory({
      localUrlToLocalPath: (localUrl) => {
        const localPath = __dirname.replace('__tests__', '__mock__')
        return `${localPath}/files${localUrl.slice(7)}`
      },
    })

    return render(file).then(vfile => {
      expect(vfile.messages).toEqual([])
      expect(r(vfile.contents)).toBe(html)
    })
  })

  test('copies local images with replacement array', () => {
    const file = `![](/foobar/ok.png)`
    const html = `<p><img src="foo/bar.png"></p>`

    const render = renderFactory({
      localUrlToLocalPath: [
        '/foobar',
        testFilesDir,
      ],
    })

    return render(file).then(vfile => {
      expect(vfile.messages).toEqual([])
      expect(r(vfile.contents)).toBe(html)
    })
  })

  test('copies local SVG with replacement array', () => {
    const file = `![](/foobar/test.svg)`
    const html = `<p><img src="foo/bar.svg"></p>`

    const render = renderFactory({
      localUrlToLocalPath: [
        '/foobar',
        testFilesDir,
      ],
    })

    return render(file).then(vfile => {
      expect(vfile.messages).toEqual([])
      expect(r(vfile.contents)).toBe(html)
    })
  })

  test('copies local images with replacement array', () => {
    const file = `![](/foobar/ok.png)`
    const html = `<p><img src="foo/bar.png"></p>`

    const render = renderFactory({
      localUrlToLocalPath: [
        '/foobar',
        testFilesDir,
      ],
    })

    return render(file).then(vfile => {
      expect(vfile.messages).toEqual([])
      expect(r(vfile.contents)).toBe(html)
    })
  })

  test('does not copy bad local SVG', () => {
    const file = `![](/foobar/bad.svg)`
    const html = `<p><img src="/foobar/bad.svg"></p>`

    const render = renderFactory({
      localUrlToLocalPath: [
        '/foobar',
        testFilesDir,
      ],
    })

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toMatch('mime type, not SVG either')
      expect(r(vfile.contents)).toBe(html)
    })
  })

  test('ignores nonexistent local files', () => {
    const file = `![](/does-not-exist)`
    const html = `<p><img src="/does-not-exist"></p>`

    const render = renderFactory({
      localUrlToLocalPath: ['', `${__dirname}`],
    })

    return render(file).then(vfile => {
      // XXX probably okay but a bit dirty
      expect(firstMsg(vfile)).toMatch('ENOENT: no such file or directory, open ')
      expect(r(vfile.contents)).toBe(html)
    })
  })

  test('ignores empty local files', () => {
    const file = `![](/empty)`
    const html = `<p><img src="/empty"></p>`

    const render = renderFactory({
      localUrlToLocalPath: ['', testFilesDir],
    })

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toMatch('Empty file: ')
      expect(r(vfile.contents)).toBe(html)
    })
  })

  test('does not copy non-image files', () => {
    const file = `![](/foobar/wrong-mime.zip)`
    const html = `<p><img src="/foobar/wrong-mime.zip"></p>`

    const render = renderFactory({
      localUrlToLocalPath: [
        '/foobar',
        testFilesDir,
      ],
    })

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toMatch('is not an image/ type')
      expect(r(vfile.contents)).toBe(html)
    })
  })

  test('does not copy dangerous local absolute URLs', () => {
    const file = `![](/../../../etc/shadow)`
    const html = `<p><img src="/../../../etc/shadow"></p>`
    const render = renderFactory({
      localUrlToLocalPath: (localUrl) => {
        const localPath = __dirname.replace('__tests__', '__mock__')
        return `${localPath}/files${localUrl.slice(7)}`
      },
    })

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toMatch('Dangerous absolute image URL detected')
      expect(r(vfile.contents)).toBe(html)
    })
  })

  test('reports 404', () => {
    const file = `![](http://localhost:27273/404/notfound)`
    const html = `<p><img src="http://localhost:27273/404/notfound"></p>`

    const render = renderFactory()

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toBe(
        'Received HTTP404 for: http://localhost:27273/404/notfound')
      expect(vfile.contents).toBe(html)
    })
  })

  test('default bad status code', () => {
    const defaultImagePath = 'default.png'
    const file = `![](http://localhost:27273/noimage.png)`
    const html = `<p><img src="${downloadDestination}/${defaultImagePath}"></p>`

    const render = renderFactory({
      defaultImagePath,
      defaultOn: {
        statusCode: true,
      },
    })

    return render(file).then(vfile => {
      expect(vfile.contents).toBe(html)
    })
  })

  test('reports DNS resolution issues', () => {
    const file = `![](http://doesnotexist-kqwtkk78.com)`
    const html = `<p><img src="http://doesnotexist-kqwtkk78.com"></p>`

    const render = renderFactory({httpRequestTimeout: 500})

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toContain(
        'getaddrinfo ENOTFOUND doesnotexist-kqwtkk78.com')
      expect(vfile.contents).toBe(html)
    })
  })

  test('report connection errors', () => {
    const file = `![](http://example.com:32764/bad)`
    const html = `<p><img src="http://example.com:32764/bad"></p>`

    const render = renderFactory({httpRequestTimeout: 500})

    return render(file).then(vfile => {
      expect(firstMsg(vfile)).toBe(
        'Request for http://example.com:32764/bad timed out')
      expect(vfile.contents).toBe(html)
    })
  }, 600)

  test('does not download large chunked streams', async () => {
    const file = `![](http://localhost:27273/stream-bomb)`
    const html = `<p><img src="http://localhost:27273/stream-bomb"></p>`

    const render = renderFactory()

    const error = 'File at http://localhost:27273/stream-bomb weighs more than 1000000'

    await noDownloads(async () => {
      const vfile = await render(file)
      expect(vfile.data.imageDir).toBe(undefined)
      expect(vfile.contents).toBe(html)
      expect(vfile.messages.length).toBe(1)
      expect(firstMsg(vfile)).toBe(error)
    })
  })

  test('rejects empty files', async () => {
    const file = `![](http://localhost:27273/empty)`
    const html = `<p><img src="http://localhost:27273/empty"></p>`

    const render = renderFactory()

    const error = 'File at http://localhost:27273/empty is empty'

    await noDownloads(async () => {
      const vfile = await render(file)
      expect(vfile.data.imageDir).toBe(undefined)
      expect(vfile.contents).toBe(html)
      expect(vfile.messages.length).toBe(1)
      expect(firstMsg(vfile)).toBe(error)
    })
  })

  test('does not download many times the same image', async () => {
    const file = dedent`
      ![](http://localhost:27273/ok.png)
      ![](http://localhost:27273/ok.png)
    `

    const html = new RegExp(dedent`
      <p><img src="([^"]+)">
      <img src="([^"]+)"></p>
    `)

    const render = renderFactory()

    return render(file)
      .then(vfile => {
        expect(vfile.messages.length).toBe(0)
        const match = vfile.contents.match(html)
        expect(match).toBeTruthy()
        expect(match[1]).toBe(match[2])
        const downloadedFiles = fs.readdirSync(vfile.data.imageDir)
        expect(downloadedFiles.length).toBe(1)
      })
  })
})
