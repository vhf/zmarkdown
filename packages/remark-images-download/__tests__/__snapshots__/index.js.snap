// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skips bigger images and reports 1`] = `"File at http://example.com/too-big.png weighs 99999999, max size is 1000000"`;

exports[`skips when directory reaches size limit 1`] = `"Cannot download http://example.com/30percent.png because destination directory reached size limit"`;

exports[`skips wrong mimes and reports 1`] = `"Content-Type of http://example.com/wrong-mime.png is not of image/ type"`;
